### PR TITLE
chore: Improves schema Description / MarkdownDescription

### PR DESCRIPTION
**Pull request by lantoli**
## Description

Improves schema Description / MarkdownDescription.
- Unit tests to verify that  both Description and MarkdownDescription are defined and have the same value.
- Runtime update to Description so only MarkdownDescription is needed.

Implementation details:
- Runtime changes are done with reflection as there are not setter methods so interfaces can not be used. 
- An alternative would be to have specific code for each underlaying struct but that's very verbose, for more info see https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2605/commits/94c7ef63588775511cc1bcf823c121b46c62c873
- However tests can be done with interfaces.
- Also a good thing of doing it in different ways is that bugs in a implementation style can be caught by the other one. 

Link to any related issue(s): CLOUDP-274028

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals I have added appropriate changelog entries.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/common/conversion/schema_description.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		if v.Description == "" {
+			v.Description = v.MarkdownDescription
+		} else if v.MarkdownDescription == "" {
+			v.MarkdownDescription = v.Description
+		}
+		for i := range v.NestedObject.Attributes {
+			v.NestedObject.Attributes[i] = updateAttribute(v.NestedObject.Attributes[i])
+		}
+		return v
+	}
```
Comment (EspenAlbert) [inline internal/common/conversion/schema_description.go:64]: default case. Raise an error?

File: internal/common/conversion/schema_description.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+func UpdateDSSchemaDescription(s *dsschema.Schema) {
+	updateAttr(s)
+}
+
+func updateAttr(attr any) {
+	ptr := reflect.ValueOf(attr)
+	if ptr.Kind() != reflect.Ptr {
+		panic("not ptr, please fix caller")
+	}
+	v := ptr.Elem()
```
Comment (lantoli) [inline internal/common/conversion/schema_description.go:20]: they're short var names but it's in line with conventional naming when doing reflection

File: internal/common/conversion/schema_description.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	if v.Kind() != reflect.Struct {
+		panic("not struct, please fix caller")
```
Comment (lantoli) [inline internal/common/conversion/schema_description.go:22]: added multiple panic to make sure it works as expected. 

These panics will be caught in unit tests so they won't make it to master, e.g. in TestResourceSchemas

File: internal/common/conversion/schema_description.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		!fMDDescr.CanSet() || fMDDescr.Kind() != reflect.String {
+		panic("invalid desc fields, please fix caller")
+	}
+	strDescr, strMDDescr := fDescr.String(), fMDDescr.String()
+	if strDescr == "" && strMDDescr != "" {
+		fDescr.SetString(fMDDescr.String())
+		return
+	}
+	if strMDDescr == "" && strDescr != "" {
+		fMDDescr.SetString(fDescr.String())
```
Comment (lantoli) [inline internal/common/conversion/schema_description.go:48]: this is to accommodate cases like "timeout" where only Description is filled

File: internal/common/conversion/schema_description.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	if f.Kind() != reflect.Map {
+		panic("not map, please fix caller: " + mapName)
+	}
+	for _, k := range f.MapKeys() {
+		v := f.MapIndex(k).Elem()
+		newPtr := reflect.New(v.Type())
+		newPtr.Elem().Set(v)
+		updateAttr(newPtr.Interface())
+		f.SetMapIndex(k, newPtr.Elem())
```
Comment (lantoli) [inline internal/common/conversion/schema_description.go:63]: maps are not addressable so a new element is created and udpated

File: internal/common/conversion/schema_description.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package conversion
```
Comment (EspenAlbert) [inline internal/common/conversion/schema_description.go:1]: curious if we could unit test this, but might not be worth it?

Comment (lantoli) [reply internal/common/conversion/schema_description.go:1]: we could, although TestResourceSchemas is testing it indirectly

**Review by EspenAlbert — APPROVED** [review]
Great work!
Almost net negative PR 😅  +165 −143

File: internal/provider/provider_test.go [hunk 7]
```diff
@@ -73,3 +79,54 @@ func TestDataSourceSchemas(t *testing.T) {
...
+		validateBlock(name+"."+k, v, diagnostics)
+	}
+}
+
+type descriptor interface {
+	GetDescription() string
+	GetMarkdownDescription() string
+}
+
+func checkDescriptor(name string, d descriptor, diagnostics *diag.Diagnostics) {
```
Comment (EspenAlbert) [inline internal/provider/provider_test.go:125]: Nice!

Comment (lantoli) [reply internal/provider/provider_test.go:125]: > Almost net negative PR 😅 +165 −143

it'll be negative in the future when more schemas are added ;-)

File: internal/common/conversion/schema_description.go (hunk 6 above)
Comment (lantoli) [reply internal/common/conversion/schema_description.go:1]: @EspenAlbert  at the end I created some unit tests here https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2605/commits/318c46cfab21a43b42bd444a66a6573dcb359a59

**Comment by lantoli** [comment]
no doc review needed as it's only deleting duplicated doc

**Review by carriecwk — COMMENTED** [review]
LGTM

**Pull request merged**